pkg/controllers/config/cloudinfo.go

```diff
@@ -24,14 +26,38 @@ type clients struct {
 
 var ci *CloudInfo
 
-// getCloudInfo fetches and caches metadata from openstack
-func getCloudInfo() (*CloudInfo, error) {
+func isMultiAZDeployment() (bool, error) {
 	var err error
 
 	if ci != nil {
-		return ci, nil
+		ci, err = getCloudInfo()
+		if err != nil {
+			return false, fmt.Errorf("couldn't collect info about cloud availability zones: %w", err)
+		}
+	}
+
+	// TODO: Remove these debug logs
+	klog.V(2).Infof("found cloud info: %+v", ci)
+	klog.V(2).Infof("found compute zones: %+v", ci.ComputeZones)
+	klog.V(2).Infof("found volume zones: %+v", ci.ComputeZones)
+
+	// There should be at least one Compute Zone and one Volume Zone. If there
```

Comment:
Is that really a given? Let's be conservative and not assume that clouds have at least one AZ.

I think we shouldn't enter this condition anyway now that you've fixed the caching issue. On first call getCloudInfo() was initializing the global `ci` to a non-nil value and before exiting with an error, then returned early because the value was already initialized. You've now moved the initialization of `ci` out of getCloudInfo() and it remains nil when getCloudInfo() returns an error. We should be good I think.

Reply:
For nova it's guaranteed: there will always be a `nova` AZ. I'm pretty sure cinder is the same. I've fixed the caching issue but the fix relies on this assumption. You'll see here that I am still caching `ci`, only in this function instead of in `getCloudInfo()` and only when we find a non-zero number of AZs in nova and cinder. If we don't want to rely on this assumption, I think we either need to remove caching entirely or drop the whole feature.